src_c/event.c

```diff
@@ -928,14 +971,37 @@ dict_from_event(SDL_Event *event)
     long gain;
     long state;
 
-    /* check if a proxy event or userevent was posted */
-    if (event->type >= PGPOST_EVENTBEGIN && event->user.code == USEROBJ_CHECK)
-        return (PyObject *)event->user.data1;
-
     dict = PyDict_New();
     if (!dict)
         return NULL;
 
+    /* check if a proxy event or userevent was posted */
+    if (event->type >= PGPOST_EVENTBEGIN) {
+        int is_freed;
+        pgEventDictProxy *dict_proxy = (pgEventDictProxy *)event->user.data1;
+        if (!dict_proxy) {
+            /* the field being NULL implies empty dict */
+            return dict;
+        }
```

Comment:
Nit: As a slight optimization, we could move the `dict = PyDict_New();` down below the `if (event->type >= PGPOST_EVENTBEGIN) {` and then replace `return dict;` with `return PyDict_New();`).

It would avoid an unnecessary allocation when we are processing a dict proxy and would have the same speed otherwise (technically, it would be one `if`-less, but I am not sure we are in that hot of a code path).